### DISCUSSION TITLE
localStorage is not defined

### DISCUSSION THREAD

**Discussion by Himanshu-holmes** [vercel/next.js#49131] [Help] [answered]
### Summary

import Image from 'next/image'
import { Inter } from 'next/font/google'
import styles from '@/styles/Home.module.css'
import 'bootstrap/dist/css/bootstrap.min.css'
import { useState, useEffect } from 'react'
import femaleProfile from '../../public/femaleProfile.jpg';
import maleProfile from '../../public/maleProfile.jpg'
const inter = Inter({ subsets: ['latin'] })

export default function App () {
  const [selectedTeam, setTeam] = useState(JSON.parse(
    localStorage.getItem('selectedTeam'))  "TeamB");
  const [employees, setEmployees] = useState(JSON.parse(localStorage.getItem('employeeList'))  [{
    id: 1,
    fullName: "Bob Jones",
    designation: "JavaScript Developer",
    gender: "male",
    teamName: "TeamA"
  },
  {
    id: 2,
    fullName: "Jill Bailey",
    designation: "Node Developer",
    gender: "female",
    teamName: "TeamA"
  },
  {
    id: 3,
    fullName: "Gail Shepherd",
    designation: "Java Developer",
    gender: "female",
    teamName: "TeamA"
  },
  {
    id: 4,
    fullName: "Sam Reynolds",
    designation: "React Developer",
    gender: "male",
    teamName: "TeamB"
  },
  {
    id: 5,
    fullName: "David Henry",
    designation: "DotNet Developer",
    gender: "male",
    teamName: "TeamB"
  },
  {
    id: 6,
    fullName: "Sarah Blake",
    designation: "SQL Server DBA",
    gender: "female",
    teamName: "TeamB"
  },
  {
    id: 7,
    fullName: "James Bennet",
    designation: "Angular Developer",
    gender: "male",
    teamName: "TeamC"
  },
  {
    id: 8,
    fullName: "Jessica Faye",
    designation: "API Developer",
    gender: "female",
    teamName: "TeamC"
  },
  {
    id: 9,
    fullName: "Lita Stone",
    designation: "C++ Developer",
    gender: "female",
    teamName: "TeamC"
  },
         // I am getting Server Error
ReferenceError: localStorage is not defined           help me how to fixt it in nextjs

**Comment by icyJoseph** [comment]
Hi,

You are making a common mistake, which is accessing the Window/DOM during SSR.

Most people have learned React in the context of a purely CSR driven application. In Client Side Rendered apps, the window/DOM is always present, so you can just access it.

When you move onto SSR or a hybrid framework such as Next.js, you have to be careful when accessing the window/DOM, because your code might be running on the server, and APIs such as LocalStorage, are not present there. 

For the case of localStorage, what you can do is, either bring in `useEffect`, and read from the localStorage after the first render pass, updating `useState` within the effect, or you could delay the component mounting, something like:


```jsx
function Parent() {
  const [showChild, setShowChild] = useState(false);
  
  // Wait until after client-side hydration to show
  useEffect(() => {
    setShowChild(true);
  }, []);
  
  if (!showChild) {
    // You can show some kind of placeholder UI here
    return null;
  }

  return <ChildThatReadsLocalStorage {...props} />;
}

function ChildThatReadsLocalStorage(props) {
  // read the local storage here
}
```

In general, inside a use-effect and client interaction handlers, you should be able to assume that the window is present, otherwise, you gotta use `typeof window !== 'undefined'`, which checks for the presence of the Window object on the current JavaScript context. If the window is undefined, we are in SSR.

You might be tempted to use that in `useState` either with a ternary or lazy initialization, but that'd potentially trigger something called divergent HTML, where the the SSR and CSR rendered HTML are not the same. That's because SSR would not read the localStorage, but the hydration step in CSR does, and if there's data there, it might lead to different results than what the server generated.

**Reply by Himanshu-holmes** [reply]
can you show through the above program

**Reply by icyJoseph** [reply]
No, mostly because the code seems incorrect to begin with. Helping here with a code-copy-paste might also do more harm than good. 

I can give you some pseudo-code, but I think a simple google search for "localstorage ssr react" or next.js should yield pretty good results.

The important bit here is to correctly understand:
-  why you cannot access local-storage as you wanted to,
-  and how to use the React toolbox to get around the issue. 

At its core, you want to defer reading/writing to local storage, to the client side.

```jsx
// pseudo-code, won't work if copy pasted
const useLocalStorage  = (key, initialValue) => {
  const [value, setValue] = useState(initialValue);
  
  useEffect(() => {
     // this process might fail if local storage is somehow not available
     const currentValue = parse(localStorage.getItem(key)) // parse definition is up to you
     setValue(currentValue);
  }, [key])

  // some hooks you might find online will create a setItem hook here, and such
  // this pseudo-code only reads the value when key changes

  return value
}
```

So the intention here would be to trigger an immediate re-render once in the client, which returns a given value if found. 

You need to account for a sensible default UI for the SSR process. 

Some literature:

- https://react.dev/reference/react/useEffect#displaying-different-content-on-the-server-and-the-client
- https://stackoverflow.com/questions/16427636/check-if-localstorage-is-available (read through it, it is important to not assume the presence of localStorage, it might be disabled for various reasons)

Happy to answer questions as you try to figure this out.

**Reply by Himanshu-holmes** [reply]
plz help me I am learning react but as you know I amde my app in 
nextjs

**Reply by Himanshu-holmes** [reply]
useEffect(() => {localStorage.setItem('employeeList',JSON.stringify(employees))},[employees]);

  useEffect(() => {localStorage.setItem('selectedTeam',JSON.stringify(selectedTeam))
},[selectedTeam]);

**Reply by Himanshu-holmes** [reply]
I tried this `const [selectedTeam, setTeam] = useState(false);

  useEffect(()=> {
    setTeam(true);
  }, []);
  if(setTeam) {
    useEffect(() => {localStorage.setItem('selectedTeam',JSON.stringify(selectedTeam))
  },[selectedTeam]);
  }else {
    JSON.parse(
      localStorage.getItem('selectedTeam')) || "TeamB";
  }
`  //but not working

**Reply by icyJoseph** [reply]
Aha, okay, do you think you can create a https://stackblitz.com/ Next.js example, with your use case, I can then modify it and share it back here.

**Reply by Himanshu-holmes** [reply]
https://stackblitz.com/edit/nextjs-zjdlus?file=pages/index.js

**Reply by Himanshu-holmes** [reply]
![maleProfile](https://user-images.githubusercontent.com/78066262/235964064-b63c3b6e-7690-43a7-a1c0-64973f4511fb.jpg)

**Reply by Himanshu-holmes** [reply]
![femaleProfile](https://user-images.githubusercontent.com/78066262/235964212-b9a2b577-f21c-4fb7-b1b3-d389d61d8bca.jpg)

**Reply by icyJoseph** [reply, marked as answer]
Alright, sorry I was out of office. 

I've rearranged your code so that it barely works.

https://stackblitz.com/edit/nextjs-ce36vm?file=pages%2Findex.js

The main thing is that I made use of `getter` functions to collect the value from the local storage, or otherwise use a default.

```js
const getSelectedTeamFromLocalStorage = () =>
  JSON.parse(localStorage.getItem('selectedTeam')) || 'TeamB';

const initialEmployeesList = [ /* some long list */]

const getEmployeesFromLocalStorage = () =>
  JSON.parse(localStorage.getItem('employeeList')) || initialEmployeesList;

```

And used them like:

```js
  const [selectedTeam, setTeam] = useState('TeamB');
  const [employees, setEmployees] = useState(initialEmployeesList);

  useEffect(() => {
    setTeam(getSelectedTeamFromLocalStorage());
  }, []);

  useEffect(() => {
    setEmployees(getEmployeesFromLocalStorage());
  }, []);
```

The one thing to appreciate is that there's a frame where the selected team is TeamB, even if you have saved TeamC as selected, but then immediately the use-effect runs and updates to TeamC.

> https://usehooks-ts.com/react-hook/use-isomorphic-layout-effect

I am also not sure why you store the employee list to local storage. Remember that serialisation (into string) to local storage, and reading that back into an object is an "expensive" process. Larger objects take longer time, so if you could skip that, it'd be nice.

I also made some changes to the next/image usage, which expects width and height to be pixels: https://nextjs.org/docs/api-reference/next/image#width, and are passed without the unit.

**Reply by Himanshu-holmes** [reply]
Thank you so much for giving your precious time I am glad.

**Reply by icyJoseph** [reply]
No worries. Don't forget to mark as answered, and ask further if you have doubts.

**Reply by Himanshu-holmes** [reply]
Hi I am confused due to ai and ibm news I am afraid should I learn web dev or learn something else.. guide me plz

**Reply by icyJoseph** [reply]
You should not do news-driven-development, nor learning. It is true that LLM's are really good at spitting out code, but that's not everything that's at play here.

A production app is not only code it runs, there's white-labelling, optimisation, changes in UX direction, styling, HTML semantics requirements, etc... In other words, managing a production application has many dimensions, naturally.

In any area of software development, the barriers of entry are constantly being lowered, because of all of the tools we have access to, but at the same time, it becomes clearer that the people who truly master their craft are more necessary than ever. 

People who can point out where the tooling is wrong, or that can amplify their skills through it. People who can put orchestrate these tools to build smaller building blocks of a bigger project, etc... These people know the basics in and out.

I think in particular for FrontEnd Dev, being able to manage stylesheets, theming, typography, HTML semantics, A11Y, data flow through the app, performance, resource loading, and such are still things a developer has to manage.

That being said, if I was in your position, on a learning stage, I'd also put some time into learning how to use these tools to my advantage, and also learn some Computer Science concepts.

**Reply by Himanshu-holmes** [reply]
thank you again

**Comment by YuluDuan** [comment]
hello @icyJoseph  I also have the quite same problem but I was wondering since I am using the newest version of next.js with App router and i annotate at the top with `'use client'` but I am still not able to access to the localStorage, do you have any ideas about that? many thanks!

**Reply by icyJoseph** [reply]
There's a misunderstanding (I blame tech-influencers), that client components render only client side, but that's not correct. The client components are the components we've used up till now. And they are SSR'd:

- https://github.com/reactwg/server-components/discussions/4
- https://nextjs.org/docs/getting-started/react-essentials#client-components

> Client Components enable you to add client-side interactivity to your application. In Next.js, they are [pre-rendered](https://nextjs.org/docs/app/building-your-application/rendering#component-level-client-and-server-rendering) on the server and [hydrated](https://nextjs.org/docs/app/building-your-application/rendering#component-level-client-and-server-rendering) on the client. You can think of Client Components as how components in the [Pages Router](https://nextjs.org/docs/pages) have always worked.

It doesn't change anything. You cannot access client side API's during at pre-render SSR, nor on demand SSR.

**Reply by YuluDuan** [reply]
Thank you so much for your response that's really helpful! However, I was wondering how people usually persist data for example when I refresh the page the information is still there [basically preload from and save to LocalStorage] using Next.js?

**Reply by icyJoseph** [reply]
Well, localStorage is an "external system", https://react.dev/reference/react/useEffect#connecting-to-an-external-system, and if we stick to rendering as pure, we should not interact with the localStorage at any point of the rendering, rather, we can do that in an effect or event handlers. It is even possible to use useSyncExternalStore.

So we'd need four pieces:

- an initial value, so called, sensible default
- React state, initialised to the sensible default
- an effect that runs after first render, and synchronises the state with what we found on the localStorage
- a callback that sets state, and also synchronises the localStorage with state

> Initialising `useState` to a value found in localStorage, even with lazy initialisation, or `typeof window` guards, is not proper of a pure function, which render passes ought to be - this is why we end up in this localStorage is not defined problem, and even if we did some window guards, we'd create diverging HTML between the client and the server.

```js
import { useState, useEffect } from 'react';

const FAV_NUM_KEY = '__fav_num';

const UserFavoriteNumber = ({ initial = 0 }) => {
  const [number, setNumber] = useState(initial);

  useEffect(() => {
    const favNum = Number(localStorage.getItem(FAV_NUM_KEY));
    if (isNaN(favNum)) return;

    setNumber(favNum);
  }, []);

  const saveFavoriteNumber = (e) => {
    e.preventDefault();

    const data = new FormData(e.target);
    const favNum = Number(data.get('favNum'));

    if (isNaN(favNum)) {
       // perhaps alert the user that their input was not good enough, yada yada
       return;
    }

    setNumber(favNum);
    localStorage.setItem(FAV_NUM_KEY, `${favNum}`);

    e.target.reset();
  };

  return (
    <div>
      <p>You fav number is: {number}</p>

      <form onSubmit={saveFavoriteNumber}>
        <label>
          <span>new fav number: </span>
          <input
            type="number"
            name="favNum"
            required
            placeholder="Input a number"
          />
        </label>
        <button type="submit">save number</button>
      </form>
    </div>
  );
};
```

You should try to add a few things yourself:

- Maybe not show UI until the effect has synchronised, this could be done with CSS or with conditional JSX
- Handle failure to interact with localStorage, which might not be available
- Reseting or deleting from the localStorage

**Reply by YuluDuan** [reply]
Thanks so much for your reply, that's really helpful!

**Reply by Himanshu-holmes** [reply]
Hi @icyJoseph could you plz guide me I am learning full-stack development, I have messaged you on linkedIN plz check
